src/types.rs

```diff
@@ -522,51 +534,135 @@ pub enum DateError {
         let mut source = source.to_string();
         source.retain(|f| !f.is_whitespace());
 
-        let full_date = source.parse::<NaiveDate>();
+        let mut parsed = Parsed::new();
+        match format::parse(
+            &mut parsed,
+            &source,
+            std::iter::once(Item::Fixed(Fixed::RFC3339)),
+        ) {
+            Ok(()) => (),
+            Err(error) if error.kind() == ParseErrorKind::TooShort => (),
+            Err(_) => todo!(),
+        };
 
-        if let Ok(ndate) = full_date {
-            Ok(Self {
-                year: ndate.year(),
-                month: Some(ndate.month0() as u8),
-                day: Some(ndate.day0() as u8),
-            })
-        } else if let Some(captures) = MONTH_REGEX.captures(&source) {
-            let month = (captures.name("m").unwrap()).as_str().parse::<u8>().unwrap() - 1;
-            if month > 11 {
-                Err(DateError::MonthOutOfBounds)
-            } else {
-                Ok(Self {
-                    year: (captures.name("y").unwrap()).as_str().parse().unwrap(),
-                    month: Some(month),
-                    day: None,
-                })
-            }
-        } else if let Some(captures) = YEAR_REGEX.captures(&source) {
-            Ok(Self {
-                year: (captures.name("y").unwrap()).as_str().parse().unwrap(),
-                month: None,
-                day: None,
+        let Parsed {
+            year,
+            month,
+            day,
+            hour_div_12,
+            hour_mod_12,
+            minute,
+            second,
+            nanosecond,
+            offset,
+            ..
+        } = parsed;
+        let year = year.ok_or(DateError::UnknownFormat)?;
+        let month = month.map(|month| (month - 1) as _);
+        let day = day.map(|day| (day - 1) as _);
+        let hour = hour_div_12
+            .zip(hour_mod_12)
+            .map(|(div_12, mod_12)| (div_12 * 12 + mod_12) as _);
+        let minute = minute.map(|minute| minute as _);
+        let second = second.map(|second| second as _);
+        let offset = offset
+            .map(|offset| {
+                FixedOffset::east_opt(offset).ok_or(DateError::OffsetOutOfBounds)
             })
-        } else {
-            Err(DateError::UnknownFormat)
-        }
+            .transpose()?;
+        
+        Ok(Date {
+            year,
+            month,
+            day,
+            hour,
+            minute,
+            second,
+            nanosecond,
+            offset,
+        })
     }
 }
 
 impl Date {
     /// Get a date from an integer.
     pub fn from_year(year: i32) -> Self {
-        Self { year, month: None, day: None }
+        Self {
+            year,
+            month: None,
+            day: None,
+            hour: None,
+            minute: None,
+            second: None,
+            nanosecond: None,
+            offset: None,
+        }
```

Comment:
This is a lot of `None` and it is repeated below. Would you consider [using a derived `Default` like this](https://stackoverflow.com/a/19653453/1567835) to cut down on boilerplate?